### DISCUSSION TITLE
Unrestrained use of CPU and memory

### DISCUSSION THREAD

**Discussion by ghostry** [photoprism/photoprism#3769] [Q&A]
#### 1. What is not working as documented?

Using the newly installed photoprism with docker, when scanning 90,000 pictures, the CPU will be filled up, and the memory will continue to increase until the system crashes and restarts.

#### 2. How can we reproduce it?

1. docker run .... -m 1G ...
2. scanning 90,000 pictures
3. wait

Picture types include jpg, png, gif, bmp, etc.
Mainly jpg.
Thumbnail quality is cublic

#### 3. What behavior do you expect?

Scan files in stages and don’t crash

#### 4. What could be the cause of your problem?

Not sure, there may be a lack of memory management, or a memory leak. There also seems to be a problem with the CPU usage limit.

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

no

#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number: AMD64, cpu I7 7500u

(b) Database Type & Version: MariaDB, 10.11

(c) Operating System Types & Versions: Linux, debian12 docker

(d) Browser Types & Versions:  Chrome, 

(e) Ad Blockers, Browser Plugins, and/or Firewall Software?

no

#### 7. On what kind of device is PhotoPrism installed?

(a) Device / Processor Type: cpu I7 7500u

(b) Physical Memory & Swap Space in 4GB

(c) Storage Type: Network Storage, nfs4，2.5G network port

(d) Anything else that might be helpful to know?

Hard disk read and write speed is about 140MB/s

#### 8. Do you use a Reverse Proxy, Firewall, VPN, or CDN?

no

**Comment by lastzero** [comment]
If you restrict the memory to 1 GB with Docker as shown, the expected behavior is that it kills PhotoPrism. So this works as designed and our requirements clearly say you need 4 GB plus swap. So not 1 GB. Please read the full docs and limit the number of workers (as documented) to reduce the memory usage. You can additionally disable features.

**Reply by ghostry** [reply]
This is not the point. I tried to give it 6G and it could still be used up and crash and restart. Should photoprism check how much memory there is before using it, rather than using it recklessly.

**Reply by lastzero** [reply]
It just needs additional memory while indexing depending on how large your images are. This is not under our control. Of course we could use a C library (instead of native Go code) that processes images in a memory restricted way (in chunks), but then you may be exposed to buffer overflows and all the other advantages C has to offer. Or you contribute Go code that does that. I don't have time right now because I am on vacation.
